File: 0179_largest_number/largest_number.java
class Solution {
    // Time: O(n(log n))
    // Space: O(n)
    public String largestNumber(int[] nums) {
        int n = nums.length;
        if (n == 0) {
            return "";
        }
        String[] numbersAsStrings = new String[n];
        for (int i = 0; i < n; i++) {
            numbersAsStrings[i] = String.valueOf(nums[i]);
        }
        Arrays.sort(numbersAsStrings, (a, b) -> (b + a).compareTo(a + b));
        if (numbersAsStrings[0].equals("0")) {
            return "0";
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(numbersAsStrings[i]);
        }
        return sb.toString();
    }

    public String largestNumberWithStreams(int[] nums) {
        return Arrays.stream(nums)
                     .mapToObj(Integer::toString)
                     .sorted((a, b) -> (b + a)
                     .compareTo(a + b))
                     .reduce((a, b) -> a + b)
                     .filter(str -> str.charAt(0) != '0')
                     .orElse("0");
    }
}

File: 0179_largest_number/largest_number.py
from functools import cmp_to_key
from typing import List


def compare(a, b):
    x, y = a + b, b + a
    if x == y:
        return 0
    return 1 if x > y else -1


def largest_number(nums: List[int]) -> str:
    """
    Time: O(n(log n))
    Space: O(n)
    """
    strings = [str(x) for x in nums]
    strings.sort(key=cmp_to_key(compare), reverse=True)
    return str(int(''.join(strings)))
